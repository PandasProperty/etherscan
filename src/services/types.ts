type GasPricesResult = {
  LastBlock: string;
  SafeGasPrice: string;
  ProposeGasPrice: string;
  FastGasPrice: string;
  suggestBaseFee: string;
  gasUsedRatio: string;
};

type EtherPriceResult = {
  ethbtc: string;
  ethbtc_timestamp: string;
  ethusd: string;
  ethusd_timestamp: string;
};

export type ServerResponse<T> = {
  status: string;
  message: string;
  result: T;
};

export type ServerGasPricesResponse = ServerResponse<GasPricesResult>;
export type ServerEtherPriceResponse = ServerResponse<EtherPriceResult>;
