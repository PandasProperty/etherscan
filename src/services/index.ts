import { EtherPrice, GasPrices, Status } from '@/types';
import { baseUrl } from '@/utils/constants';
import { ServerEtherPriceResponse, ServerGasPricesResponse } from './types';

export async function fetchGasPrices(abortController: AbortController): Promise<Status<GasPrices>> {
  const searchParams = new URLSearchParams();
  searchParams.set('module', 'gastracker');
  searchParams.set('action', 'gasoracle');
  searchParams.set('apikey', process.env.NEXT_PUBLIC_ETHER_SCAN_API_KEY!);
  let url = new URL(`${baseUrl}?${searchParams.toString()}`)

  console.log('url', url.href);

  try {
    const response = await fetch(url.href, {
      signal: abortController.signal,
    });
  
    const data: ServerGasPricesResponse = await response.json();

    if (data.status !== "1" || data.message !== "OK") {
      return {
        error: data.message,
        data: undefined,
      };
    }
    return {
      error: undefined,
      data: {
        safeGasPrice: Number(data.result.SafeGasPrice),
        proposeGasPrice: Number(data.result.ProposeGasPrice),
        fastGasPrice: Number(data.result.FastGasPrice)
      }
    };
  } catch (error: any) {
    if (error.name === 'AbortError') {
      return {};
    }
    return {
      error: error.message,
      data: undefined,
    };
  }
}


export async function fetchEtherPrice(abortController: AbortController): Promise<Status<EtherPrice>> {
  const searchParams = new URLSearchParams();
  searchParams.set('module', 'stats');
  searchParams.set('action', 'ethprice');
  searchParams.set('apikey', process.env.NEXT_PUBLIC_ETHER_SCAN_API_KEY!);
  let url = new URL(`${baseUrl}?${searchParams.toString()}`)

  try {
    const response = await fetch(url.href, {
      signal: abortController.signal,
    });
  
    const data: ServerEtherPriceResponse = await response.json();

    if (data.status !== "1" || data.message !== "OK") {
      return {
        error: data.message,
        data: undefined,
      };
    }
    return {
      error: undefined,
      data: {
        ethbtc: Number(data.result.ethbtc),
        ethusd: Number(data.result.ethusd),
      }
    };
  } catch (error: any) {
    if (error.name === 'AbortError') {
      return {};
    }
    return {
      error: error.message,
      data: undefined,
    };
  }
}