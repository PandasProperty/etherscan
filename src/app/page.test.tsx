import "@testing-library/jest-dom"
import { act, fireEvent, render, waitFor } from "@testing-library/react";

import services from "@/services";
import {
  DATA_ETHER_PRICE_TEST_ID,
  DATA_GAS_PRICES_TEST_ID,
  ERROR_ETHER_PRICE_TEST_ID,
  ERROR_GAS_PRICES_TEST_ID,
  LOADING_ETHER_PRICE_TEST_ID,
  LOADING_GAS_PRICES_TEST_ID,
  START_POLLING_BANNER
} from "@/tests/dataTestIds";

import Home from "./page";
import { mockStatusEtherPrice, mockStatusGasPrices } from "@/tests/mocks";

describe("Home page", () => {
  it("match initial snapshot", () => {
    const { asFragment, getByTestId } = render(<Home />);
    expect(asFragment()).toMatchSnapshot();

    expect(getByTestId(LOADING_GAS_PRICES_TEST_ID)).toBeInTheDocument();
    expect(getByTestId(LOADING_ETHER_PRICE_TEST_ID)).toBeInTheDocument();
  });

  it("close start polling banner", () => {
    const { asFragment, getByTestId, queryByTestId } = render(<Home />);
    expect(asFragment()).toMatchSnapshot();
    expect(queryByTestId(START_POLLING_BANNER)).toBeInTheDocument();

    act(() => {
      fireEvent.click(getByTestId(`${START_POLLING_BANNER}-button`));
    });

    waitFor(() => {
      expect(queryByTestId(START_POLLING_BANNER)).not.toBeInTheDocument();
    });
  });

  it("match snapshot with success data", async () => {
    const mockFetchGasPrices = jest.fn();
    mockFetchGasPrices.mockImplementation(() => new Promise(resolve => resolve(mockStatusGasPrices)));
    jest.spyOn(services, "fetchGasPrices").mockImplementation(mockFetchGasPrices);

    const mockFetchEtherPrice = jest.fn();
    mockFetchEtherPrice.mockImplementation(() => new Promise(resolve => resolve(mockStatusEtherPrice)));
    jest.spyOn(services, "fetchEtherPrice").mockImplementation(mockFetchEtherPrice);
    
    const { asFragment, queryByTestId } = render(<Home />);

    await waitFor(() => {
      expect(queryByTestId(DATA_GAS_PRICES_TEST_ID)).toBeInTheDocument();
      expect(queryByTestId(DATA_ETHER_PRICE_TEST_ID)).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it("match snapshot with error data ", async () => {
    const mockFetchGasPrices = jest.fn();
    mockFetchGasPrices.mockImplementation(() => new Promise(resolve => resolve({
      error: "Error fetching gas prices."
    })));
    jest.spyOn(services, "fetchGasPrices").mockImplementation(mockFetchGasPrices);

    const mockFetchEtherPrice = jest.fn();
    mockFetchEtherPrice.mockImplementation(() => new Promise(resolve => resolve({
      error: "Error fetching ether price."
    })));
    jest.spyOn(services, "fetchEtherPrice").mockImplementation(mockFetchEtherPrice);
    
    const { asFragment, queryByTestId } = render(<Home />);

    await waitFor(() => {
      expect(queryByTestId(ERROR_GAS_PRICES_TEST_ID)).toBeInTheDocument();
      expect(queryByTestId(ERROR_ETHER_PRICE_TEST_ID)).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
