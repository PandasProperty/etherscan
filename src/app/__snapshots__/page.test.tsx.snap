// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page close start polling banner 1`] = `
<DocumentFragment>
  <main
    class="flex min-h-screen flex-col items-center gap-12 lg:p-24 p-16"
  >
    <p
      class="text-4xl text-center"
    >
      Real time information from  
      <a
        class="font-mono font-bold"
        href="https://etherscan.io/"
      >
        etherscan.io
      </a>
    </p>
    <div
      class="w-full lg:max-w-5xl py-6 flex flex-col md:flex-row justify-around items-center md:items-start from-inherit border bg-gradient-to-b backdrop-blur-2xl border-neutral-800 bg-zinc-800/30 rounded-xl bg-zinc-800/30"
    >
      <div
        class="px-6 py-4 flex flex-col divide-y divide-slate-600"
      >
        <p
          class="text-2xl pb-4"
        >
          Gas prices
        </p>
        <div
          class="p-8 self-center"
          data-testid="LoadingGasPrices"
          role="status"
        >
          <svg
            aria-hidden="true"
            class="inline w-8 h-8 text-gray-200 animate-spin dark:text-gray-600 fill-blue-600"
            fill="none"
            viewBox="0 0 100 101"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M100 50.5908C100 78.2051 77.6142 100.591 50 100.591C22.3858 100.591 0 78.2051 0 50.5908C0 22.9766 22.3858 0.59082 50 0.59082C77.6142 0.59082 100 22.9766 100 50.5908ZM9.08144 50.5908C9.08144 73.1895 27.4013 91.5094 50 91.5094C72.5987 91.5094 90.9186 73.1895 90.9186 50.5908C90.9186 27.9921 72.5987 9.67226 50 9.67226C27.4013 9.67226 9.08144 27.9921 9.08144 50.5908Z"
              fill="currentColor"
            />
            <path
              d="M93.9676 39.0409C96.393 38.4038 97.8624 35.9116 97.0079 33.5539C95.2932 28.8227 92.871 24.3692 89.8167 20.348C85.8452 15.1192 80.8826 10.7238 75.2124 7.41289C69.5422 4.10194 63.2754 1.94025 56.7698 1.05124C51.7666 0.367541 46.6976 0.446843 41.7345 1.27873C39.2613 1.69328 37.813 4.19778 38.4501 6.62326C39.0873 9.04874 41.5694 10.4717 44.0505 10.1071C47.8511 9.54855 51.7191 9.52689 55.5402 10.0491C60.8642 10.7766 65.9928 12.5457 70.6331 15.2552C75.2735 17.9648 79.3347 21.5619 82.5849 25.841C84.9175 28.9121 86.7997 32.2913 88.1811 35.8758C89.083 38.2158 91.5421 39.6781 93.9676 39.0409Z"
              fill="currentFill"
            />
          </svg>
        </div>
      </div>
      <div
        class="px-6 py-4 flex flex-col divide-y divide-slate-600"
      >
        <p
          class="text-2xl pb-4"
        >
          Ether price
        </p>
        <div
          class="p-8 self-center"
          data-testid="LoadingEtherPrice"
          role="status"
        >
          <svg
            aria-hidden="true"
            class="inline w-8 h-8 text-gray-200 animate-spin dark:text-gray-600 fill-blue-600"
            fill="none"
            viewBox="0 0 100 101"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M100 50.5908C100 78.2051 77.6142 100.591 50 100.591C22.3858 100.591 0 78.2051 0 50.5908C0 22.9766 22.3858 0.59082 50 0.59082C77.6142 0.59082 100 22.9766 100 50.5908ZM9.08144 50.5908C9.08144 73.1895 27.4013 91.5094 50 91.5094C72.5987 91.5094 90.9186 73.1895 90.9186 50.5908C90.9186 27.9921 72.5987 9.67226 50 9.67226C27.4013 9.67226 9.08144 27.9921 9.08144 50.5908Z"
              fill="currentColor"
            />
            <path
              d="M93.9676 39.0409C96.393 38.4038 97.8624 35.9116 97.0079 33.5539C95.2932 28.8227 92.871 24.3692 89.8167 20.348C85.8452 15.1192 80.8826 10.7238 75.2124 7.41289C69.5422 4.10194 63.2754 1.94025 56.7698 1.05124C51.7666 0.367541 46.6976 0.446843 41.7345 1.27873C39.2613 1.69328 37.813 4.19778 38.4501 6.62326C39.0873 9.04874 41.5694 10.4717 44.0505 10.1071C47.8511 9.54855 51.7191 9.52689 55.5402 10.0491C60.8642 10.7766 65.9928 12.5457 70.6331 15.2552C75.2735 17.9648 79.3347 21.5619 82.5849 25.841C84.9175 28.9121 86.7997 32.2913 88.1811 35.8758C89.083 38.2158 91.5421 39.6781 93.9676 39.0409Z"
              fill="currentFill"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      class="fixed lg:bottom-4 bottom-8 right-4 max-w-[80%] transition-opacity duration-5000 ease-in-out opacity-100"
    >
      <div
        class="flex items-center w-full max-w-lg p-4 text-gray-500 bg-white rounded-lg shadow dark:text-gray-400 dark:bg-gray-800"
        data-testid="StartPollingBanner"
      >
        <div
          class="inline-flex items-center justify-center flex-shrink-0 w-8 h-8  text-sm text-green-800 border border-green-300 rounded-lg bg-green-50 dark:bg-gray-800 dark:text-green-400 dark:border-green-800"
          role="alert"
        >
          <svg
            aria-hidden="true"
            class="flex-shrink-0 inline w-4 h-4"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10 .5a9.5 9.5 0 1 0 9.5 9.5A9.51 9.51 0 0 0 10 .5ZM9.5 4a1.5 1.5 0 1 1 0 3 1.5 1.5 0 0 1 0-3ZM12 15H8a1 1 0 0 1 0-2h1v-3H8a1 1 0 0 1 0-2h2a1 1 0 0 1 1 1v4h1a1 1 0 0 1 0 2Z"
            />
          </svg>
        </div>
        <div
          class="ms-3 text-sm font-normal"
          data-testid="StartPollingBanner-text"
        >
          Data will be polled from the server every 10 seconds.
        </div>
        <button
          aria-label="Close"
          class="ml-3.5 rounded-lg focus:ring-2 focus:ring-gray-300hover:b g-gray-100 inline-flex items-center justify-center h-8 w-8 text-gray-500 hover:text-white bg-gray-800 hover:bg-gray-700"
          data-dismiss-target="#toast-danger"
          data-testid="StartPollingBanner-button"
          type="button"
        >
          <span
            class="sr-only"
          >
            Close
          </span>
          <svg
            aria-hidden="true"
            class="w-3 h-3"
            fill="none"
            viewBox="0 0 14 14"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m1 1 6 6m0 0 6 6M7 7l6-6M7 7l-6 6"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </button>
      </div>
    </div>
  </main>
</DocumentFragment>
`;

exports[`Home page match initial snapshot 1`] = `
<DocumentFragment>
  <main
    class="flex min-h-screen flex-col items-center gap-12 lg:p-24 p-16"
  >
    <p
      class="text-4xl text-center"
    >
      Real time information from  
      <a
        class="font-mono font-bold"
        href="https://etherscan.io/"
      >
        etherscan.io
      </a>
    </p>
    <div
      class="w-full lg:max-w-5xl py-6 flex flex-col md:flex-row justify-around items-center md:items-start from-inherit border bg-gradient-to-b backdrop-blur-2xl border-neutral-800 bg-zinc-800/30 rounded-xl bg-zinc-800/30"
    >
      <div
        class="px-6 py-4 flex flex-col divide-y divide-slate-600"
      >
        <p
          class="text-2xl pb-4"
        >
          Gas prices
        </p>
        <div
          class="p-8 self-center"
          data-testid="LoadingGasPrices"
          role="status"
        >
          <svg
            aria-hidden="true"
            class="inline w-8 h-8 text-gray-200 animate-spin dark:text-gray-600 fill-blue-600"
            fill="none"
            viewBox="0 0 100 101"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M100 50.5908C100 78.2051 77.6142 100.591 50 100.591C22.3858 100.591 0 78.2051 0 50.5908C0 22.9766 22.3858 0.59082 50 0.59082C77.6142 0.59082 100 22.9766 100 50.5908ZM9.08144 50.5908C9.08144 73.1895 27.4013 91.5094 50 91.5094C72.5987 91.5094 90.9186 73.1895 90.9186 50.5908C90.9186 27.9921 72.5987 9.67226 50 9.67226C27.4013 9.67226 9.08144 27.9921 9.08144 50.5908Z"
              fill="currentColor"
            />
            <path
              d="M93.9676 39.0409C96.393 38.4038 97.8624 35.9116 97.0079 33.5539C95.2932 28.8227 92.871 24.3692 89.8167 20.348C85.8452 15.1192 80.8826 10.7238 75.2124 7.41289C69.5422 4.10194 63.2754 1.94025 56.7698 1.05124C51.7666 0.367541 46.6976 0.446843 41.7345 1.27873C39.2613 1.69328 37.813 4.19778 38.4501 6.62326C39.0873 9.04874 41.5694 10.4717 44.0505 10.1071C47.8511 9.54855 51.7191 9.52689 55.5402 10.0491C60.8642 10.7766 65.9928 12.5457 70.6331 15.2552C75.2735 17.9648 79.3347 21.5619 82.5849 25.841C84.9175 28.9121 86.7997 32.2913 88.1811 35.8758C89.083 38.2158 91.5421 39.6781 93.9676 39.0409Z"
              fill="currentFill"
            />
          </svg>
        </div>
      </div>
      <div
        class="px-6 py-4 flex flex-col divide-y divide-slate-600"
      >
        <p
          class="text-2xl pb-4"
        >
          Ether price
        </p>
        <div
          class="p-8 self-center"
          data-testid="LoadingEtherPrice"
          role="status"
        >
          <svg
            aria-hidden="true"
            class="inline w-8 h-8 text-gray-200 animate-spin dark:text-gray-600 fill-blue-600"
            fill="none"
            viewBox="0 0 100 101"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M100 50.5908C100 78.2051 77.6142 100.591 50 100.591C22.3858 100.591 0 78.2051 0 50.5908C0 22.9766 22.3858 0.59082 50 0.59082C77.6142 0.59082 100 22.9766 100 50.5908ZM9.08144 50.5908C9.08144 73.1895 27.4013 91.5094 50 91.5094C72.5987 91.5094 90.9186 73.1895 90.9186 50.5908C90.9186 27.9921 72.5987 9.67226 50 9.67226C27.4013 9.67226 9.08144 27.9921 9.08144 50.5908Z"
              fill="currentColor"
            />
            <path
              d="M93.9676 39.0409C96.393 38.4038 97.8624 35.9116 97.0079 33.5539C95.2932 28.8227 92.871 24.3692 89.8167 20.348C85.8452 15.1192 80.8826 10.7238 75.2124 7.41289C69.5422 4.10194 63.2754 1.94025 56.7698 1.05124C51.7666 0.367541 46.6976 0.446843 41.7345 1.27873C39.2613 1.69328 37.813 4.19778 38.4501 6.62326C39.0873 9.04874 41.5694 10.4717 44.0505 10.1071C47.8511 9.54855 51.7191 9.52689 55.5402 10.0491C60.8642 10.7766 65.9928 12.5457 70.6331 15.2552C75.2735 17.9648 79.3347 21.5619 82.5849 25.841C84.9175 28.9121 86.7997 32.2913 88.1811 35.8758C89.083 38.2158 91.5421 39.6781 93.9676 39.0409Z"
              fill="currentFill"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      class="fixed lg:bottom-4 bottom-8 right-4 max-w-[80%] transition-opacity duration-5000 ease-in-out opacity-100"
    >
      <div
        class="flex items-center w-full max-w-lg p-4 text-gray-500 bg-white rounded-lg shadow dark:text-gray-400 dark:bg-gray-800"
        data-testid="StartPollingBanner"
      >
        <div
          class="inline-flex items-center justify-center flex-shrink-0 w-8 h-8  text-sm text-green-800 border border-green-300 rounded-lg bg-green-50 dark:bg-gray-800 dark:text-green-400 dark:border-green-800"
          role="alert"
        >
          <svg
            aria-hidden="true"
            class="flex-shrink-0 inline w-4 h-4"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10 .5a9.5 9.5 0 1 0 9.5 9.5A9.51 9.51 0 0 0 10 .5ZM9.5 4a1.5 1.5 0 1 1 0 3 1.5 1.5 0 0 1 0-3ZM12 15H8a1 1 0 0 1 0-2h1v-3H8a1 1 0 0 1 0-2h2a1 1 0 0 1 1 1v4h1a1 1 0 0 1 0 2Z"
            />
          </svg>
        </div>
        <div
          class="ms-3 text-sm font-normal"
          data-testid="StartPollingBanner-text"
        >
          Data will be polled from the server every 10 seconds.
        </div>
        <button
          aria-label="Close"
          class="ml-3.5 rounded-lg focus:ring-2 focus:ring-gray-300hover:b g-gray-100 inline-flex items-center justify-center h-8 w-8 text-gray-500 hover:text-white bg-gray-800 hover:bg-gray-700"
          data-dismiss-target="#toast-danger"
          data-testid="StartPollingBanner-button"
          type="button"
        >
          <span
            class="sr-only"
          >
            Close
          </span>
          <svg
            aria-hidden="true"
            class="w-3 h-3"
            fill="none"
            viewBox="0 0 14 14"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m1 1 6 6m0 0 6 6M7 7l6-6M7 7l-6 6"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </button>
      </div>
    </div>
  </main>
</DocumentFragment>
`;

exports[`Home page match snapshot with error data  1`] = `
<DocumentFragment>
  <main
    class="flex min-h-screen flex-col items-center gap-12 lg:p-24 p-16"
  >
    <p
      class="text-4xl text-center"
    >
      Real time information from  
      <a
        class="font-mono font-bold"
        href="https://etherscan.io/"
      >
        etherscan.io
      </a>
    </p>
    <div
      class="w-full lg:max-w-5xl py-6 flex flex-col md:flex-row justify-around items-center md:items-start from-inherit border bg-gradient-to-b backdrop-blur-2xl border-neutral-800 bg-zinc-800/30 rounded-xl bg-zinc-800/30"
    >
      <div
        class="px-6 py-4 flex flex-col divide-y divide-slate-600"
      >
        <p
          class="text-2xl pb-4"
        >
          Gas prices
        </p>
        <div
          class="pt-4"
        >
          <div
            class="flex items-center w-full max-w-lg p-4 text-gray-500 bg-white rounded-lg shadow dark:text-gray-400 dark:bg-gray-800"
            data-testid="ErrorGasPrices"
          >
            <div
              class="inline-flex items-center justify-center flex-shrink-0 w-8 h-8 text-red-500 bg-red-100 rounded-lg dark:bg-red-800 dark:text-red-200"
            >
              <svg
                aria-hidden="true"
                class="w-5 h-5"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M10 .5a9.5 9.5 0 1 0 9.5 9.5A9.51 9.51 0 0 0 10 .5Zm3.707 11.793a1 1 0 1 1-1.414 1.414L10 11.414l-2.293 2.293a1 1 0 0 1-1.414-1.414L8.586 10 6.293 7.707a1 1 0 0 1 1.414-1.414L10 8.586l2.293-2.293a1 1 0 0 1 1.414 1.414L11.414 10l2.293 2.293Z"
                />
              </svg>
            </div>
            <div
              class="ms-3 text-sm font-normal"
              data-testid="ErrorGasPrices-text"
            >
              Error fetching gas prices.
            </div>
          </div>
        </div>
      </div>
      <div
        class="px-6 py-4 flex flex-col divide-y divide-slate-600"
      >
        <p
          class="text-2xl pb-4"
        >
          Ether price
        </p>
        <div
          class="pt-4"
        >
          <div
            class="flex items-center w-full max-w-lg p-4 text-gray-500 bg-white rounded-lg shadow dark:text-gray-400 dark:bg-gray-800"
            data-testid="ErrorEtherPrice"
          >
            <div
              class="inline-flex items-center justify-center flex-shrink-0 w-8 h-8 text-red-500 bg-red-100 rounded-lg dark:bg-red-800 dark:text-red-200"
            >
              <svg
                aria-hidden="true"
                class="w-5 h-5"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M10 .5a9.5 9.5 0 1 0 9.5 9.5A9.51 9.51 0 0 0 10 .5Zm3.707 11.793a1 1 0 1 1-1.414 1.414L10 11.414l-2.293 2.293a1 1 0 0 1-1.414-1.414L8.586 10 6.293 7.707a1 1 0 0 1 1.414-1.414L10 8.586l2.293-2.293a1 1 0 0 1 1.414 1.414L11.414 10l2.293 2.293Z"
                />
              </svg>
            </div>
            <div
              class="ms-3 text-sm font-normal"
              data-testid="ErrorEtherPrice-text"
            >
              Error fetching ether price.
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="fixed lg:bottom-4 bottom-8 right-4 max-w-[80%] transition-opacity duration-5000 ease-in-out opacity-100"
    >
      <div
        class="flex items-center w-full max-w-lg p-4 text-gray-500 bg-white rounded-lg shadow dark:text-gray-400 dark:bg-gray-800"
        data-testid="StartPollingBanner"
      >
        <div
          class="inline-flex items-center justify-center flex-shrink-0 w-8 h-8  text-sm text-green-800 border border-green-300 rounded-lg bg-green-50 dark:bg-gray-800 dark:text-green-400 dark:border-green-800"
          role="alert"
        >
          <svg
            aria-hidden="true"
            class="flex-shrink-0 inline w-4 h-4"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10 .5a9.5 9.5 0 1 0 9.5 9.5A9.51 9.51 0 0 0 10 .5ZM9.5 4a1.5 1.5 0 1 1 0 3 1.5 1.5 0 0 1 0-3ZM12 15H8a1 1 0 0 1 0-2h1v-3H8a1 1 0 0 1 0-2h2a1 1 0 0 1 1 1v4h1a1 1 0 0 1 0 2Z"
            />
          </svg>
        </div>
        <div
          class="ms-3 text-sm font-normal"
          data-testid="StartPollingBanner-text"
        >
          Data will be polled from the server every 10 seconds.
        </div>
        <button
          aria-label="Close"
          class="ml-3.5 rounded-lg focus:ring-2 focus:ring-gray-300hover:b g-gray-100 inline-flex items-center justify-center h-8 w-8 text-gray-500 hover:text-white bg-gray-800 hover:bg-gray-700"
          data-dismiss-target="#toast-danger"
          data-testid="StartPollingBanner-button"
          type="button"
        >
          <span
            class="sr-only"
          >
            Close
          </span>
          <svg
            aria-hidden="true"
            class="w-3 h-3"
            fill="none"
            viewBox="0 0 14 14"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m1 1 6 6m0 0 6 6M7 7l6-6M7 7l-6 6"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </button>
      </div>
    </div>
  </main>
</DocumentFragment>
`;

exports[`Home page match snapshot with success data 1`] = `
<DocumentFragment>
  <main
    class="flex min-h-screen flex-col items-center gap-12 lg:p-24 p-16"
  >
    <p
      class="text-4xl text-center"
    >
      Real time information from  
      <a
        class="font-mono font-bold"
        href="https://etherscan.io/"
      >
        etherscan.io
      </a>
    </p>
    <div
      class="w-full lg:max-w-5xl py-6 flex flex-col md:flex-row justify-around items-center md:items-start from-inherit border bg-gradient-to-b backdrop-blur-2xl border-neutral-800 bg-zinc-800/30 rounded-xl bg-zinc-800/30"
    >
      <div
        class="px-6 py-4 flex flex-col divide-y divide-slate-600"
      >
        <p
          class="text-2xl pb-4"
        >
          Gas prices
        </p>
        <div
          class="pt-4"
          data-testid="DataGasPrices"
        >
          <p
            class="py-1"
          >
            Safe gas price: 
            <b>
              4
            </b>
          </p>
          <p
            class="py-1"
          >
            Propose gas price: 
            <b>
              5
            </b>
          </p>
          <p
            class="py-1"
          >
            Fast gas price: 
            <b>
              3
            </b>
          </p>
        </div>
      </div>
      <div
        class="px-6 py-4 flex flex-col divide-y divide-slate-600"
      >
        <p
          class="text-2xl pb-4"
        >
          Ether price
        </p>
        <div
          class="pt-4"
          data-testid="DataEtherPrice"
        >
          <p
            class="py-1"
          >
            <b>
              0.048257
            </b>
             BTC
          </p>
          <p
            class="py-1"
          >
            $ 
            <b>
              3017.89
            </b>
          </p>
        </div>
      </div>
    </div>
    <div
      class="fixed lg:bottom-4 bottom-8 right-4 max-w-[80%] transition-opacity duration-5000 ease-in-out opacity-100"
    >
      <div
        class="flex items-center w-full max-w-lg p-4 text-gray-500 bg-white rounded-lg shadow dark:text-gray-400 dark:bg-gray-800"
        data-testid="StartPollingBanner"
      >
        <div
          class="inline-flex items-center justify-center flex-shrink-0 w-8 h-8  text-sm text-green-800 border border-green-300 rounded-lg bg-green-50 dark:bg-gray-800 dark:text-green-400 dark:border-green-800"
          role="alert"
        >
          <svg
            aria-hidden="true"
            class="flex-shrink-0 inline w-4 h-4"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10 .5a9.5 9.5 0 1 0 9.5 9.5A9.51 9.51 0 0 0 10 .5ZM9.5 4a1.5 1.5 0 1 1 0 3 1.5 1.5 0 0 1 0-3ZM12 15H8a1 1 0 0 1 0-2h1v-3H8a1 1 0 0 1 0-2h2a1 1 0 0 1 1 1v4h1a1 1 0 0 1 0 2Z"
            />
          </svg>
        </div>
        <div
          class="ms-3 text-sm font-normal"
          data-testid="StartPollingBanner-text"
        >
          Data will be polled from the server every 10 seconds.
        </div>
        <button
          aria-label="Close"
          class="ml-3.5 rounded-lg focus:ring-2 focus:ring-gray-300hover:b g-gray-100 inline-flex items-center justify-center h-8 w-8 text-gray-500 hover:text-white bg-gray-800 hover:bg-gray-700"
          data-dismiss-target="#toast-danger"
          data-testid="StartPollingBanner-button"
          type="button"
        >
          <span
            class="sr-only"
          >
            Close
          </span>
          <svg
            aria-hidden="true"
            class="w-3 h-3"
            fill="none"
            viewBox="0 0 14 14"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m1 1 6 6m0 0 6 6M7 7l6-6M7 7l-6 6"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </button>
      </div>
    </div>
  </main>
</DocumentFragment>
`;
